### ISSUE TITLE
handle paginated responses for tags

### ISSUE THREAD

**Issue by webratz** [deitch/docker_registry2#34]
Hey,
it seems that at least Amazon ECR is paginating responses.
The library seems not to follow the `Link` header and therefore returns only a subset of the response.
Sample URL would be `https://123456789012.dkr.ecr.eu-west-1.amazonaws.com/v2/my-imgage-name/tags/list`
The response looks sth like this:
```
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 645
Content-Type: text/plain; charset=utf-8
Date: Tue, 25 Jun 2019 09:35:44 GMT
Docker-Distribution-Api-Version: registry/2.0
Link: <https://123456789012.dkr.ecr.eu-west-1.amazonaws.com/v2/my-imgage-name/tags/list?last=LONGSTRINGWITHIDSANDSTUFF>; rel="next"

{
    "name": "my-imgage-name",
    "tags": [
        "18",
        "161",
        "138",
        "111",
// .. maaaaany more tags
        "3"
    ]
}
```
The correct thing would be to call the URl in the `Link` header unless there is no more header set.

**Comment by deitch** [comment]
Isn't that fun! Can you share a sample image name I can run against that does this?

Would the correct behaviour here be to pull them all? If so, is there an option on ECR to not paginate and return all?

Interesting that the AWS API [here](https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_ListTagsForResource.html) does not list pagination at all. They have `maxResults` for other resources, but not for that.

**Comment by webratz** [comment]
The correct behaviour would be to pull them all.
I don't think the proper solution is to just set the page size arbitrarily high, but it might be a workaround
AWS ECR seems to follow the v2 API specs defined at https://docs.docker.com/registry/spec/api/#listing-image-tags more strictly than dockerhub actually does.
Dockerhub seems to not paginate at 100 (and looks like even higher than 500 tags).

You can test that behaviour with sth like that
```bash
TOKEN=$(curl -s "https://auth.docker.io/token?service=registry.docker.io&scope=repository:jenkins/jenkins:pull"  | jq -r .token)
curl -v -s -H "Authorization: Bearer ${TOKEN}" "https://registry-1.docker.io/v2/jenkins/jenkins/tags/list?n=10" | jq -r '.tags[]' | wc -l
```
Even when setting the `n` parameter to 10 dockerhub does not paginate.
AWS ECR seems to respect that parameter, so sth like this works for my use case:
```bash
curl -sD - --user "AWS:$ECR_TOKEN" https://123456789012.dkr.ecr.eu-west-1.amazonaws.com/v2/smy-imgage-name/tags/list?n=500 | jq -r '.tags[]' | wc -l

So the easy workaround would be for now, to append `?n=500` (or some other higher number) to the request. (I assume at https://github.com/deitch/docker_registry2/blob/master/lib/registry/registry.rb#L53 )
Is this a change you would accept?

**Comment by deitch** [comment]
> Is this a change you would accept?

Yes, it is an easy workaround. That having been said, it seems the "right" approach is one of:

* have it do multiple and follow the links; some repos eventually will get to 500. Setting an arbitrary limit is like setting a timer to avoid a timing issue; eventually you hit it and wonder what happened. :-)
* add pagination to the SDK here

Exposing the pagination probably is the right thing to do, although it seems like overkill. The result could be something like:

```json
{
	"name" => "special/repo",
	"tags" => ["1.0","1.1","1.3","latest"],
        "last" => "last_index",
	"hashes" => {
		"1.0" => "abc4567",
		"1.1" => "87def23",
		"1.3" => "998adf2",
		"latest" => "998adf2"
	}
}
```

We wouldn't need the `n`, only an indicator of which the last index was in this search. The `tags()` call would then look like:

```ruby
def tags(repo,count=100,last="",withHashes = false)
```

You could just feed the value of `last` in the previous result into the next call of `tags()` as the value of the `last` arg. Conversely, if you want them all, just do `tags("myrepo", 500)`. Get the same effect as an option.

That exposes the API correctly (without making the consumer actually parse and create the header and URLs, of course), while still giving the 500 count option.

Feel like opening a PR?

**Comment by LarsFronius** [comment]
To address this issue I've created PR #35 
Let me know what you think @deitch

**Comment by deitch** [comment]
Closed via #35
